Added contest-generator

insomniacslk opened the pull request:
contest-generator is a tool used to generate a `contest` executable with
a custom set of plugins. See README.md for instructions.

Signed-off-by: Andrea Barberio <insomniac@slackware.it>

insomniacslk commented:
This is still a work in progress. I have to automatically generate cmds/contest/main.go via github actions, and validate the build.

For now this PR has:
* the implementation of `contest-generator`
* a new `cmds/contest/main.go` obtained by running `./contest-generator`
* removed `cmds/contest/server/` which is now entirely embedded in the `main.go`. This was probably used for testing, so I'm looking for usage and to replace it with the code generator

insomniacslk commented:
I see why the `server` package. I'll try to fix the e2e tests

insomniacslk commented:
Note: the listener plugin is not customizable yet, because there is no listener interface. I will add it in a later PR

rojer9-fb commented:
i'd suggest a YMAL config file instead of a custom txt, for easier future extensibility.

insomniacslk commented:
> i'd suggest a YMAL config file instead of a custom txt, for easier future extensibility.

The format could be something like:
```
[targetmanagers]
- github.com/facebookincubator/contest/plugins/targetmanagers/targetlist
- ...

[testfetchers]
- ...
```

Is this what you have in mind @rojer9-fb ?

rojer9-fb commented:
> Is this what you have in mind @rojer9-fb ?

yeah, pretty much

insomniacslk commented:
Done. The code is also a lot simpler now, thanks for the suggestion @rojer9-fb  !

rojer9-fb commented on cmds/contest-generator/main.go:
```diff
@@ -0,0 +1,134 @@
+	if len(cfg.TargetManagers) < 1 {
+		return nil, fmt.Errorf("want at least 1 targetmanager, got %d", len(cfg.TargetManagers))
+	}
+	if len(cfg.TargetManagers) < 1 {
```
all 4 conditions in this block are the same while messages are different. looks like copy-paste error.

insomniacslk replied:
done

rojer9-fb commented on cmds/contest-generator/main.go:
```diff
@@ -0,0 +1,134 @@
+	if err != nil {
+		log.Fatalf("Template parsing failed: %v", err)
+	}
+	// additional plugin names from a text file, one line per plugin import
```
this comment is outdated

insomniacslk replied:
done

insomniacslk replied:
done

rojer9-fb commented on cmds/contest-generator/main.go:
```diff
@@ -0,0 +1,134 @@
+	}
+	// additional plugin names from a text file, one line per plugin import
+	// path
+	fd, err := os.Open(*flagFromFile)
```
any reason not to use ioutil.ReadFile here?

insomniacslk replied:
It's my tendency to use `io.Reader` instead of `string` to make testing easier. However I didn't add testing to this module, and it's probably a good idea to do so at least for `parseConfig`

insomniacslk replied:
I also realized that I don't handle duplicate module names here, and I should. I'll add a check, and an optional `name` field where the user can specify an alternative import name that will result in
```
import_name "github.com/blah/duplicate_name
```

insomniacslk replied:
done

rojer9-fb commented on cmds/contest-generator/README.md:
```diff
@@ -0,0 +1,63 @@
+your own `main.go` program, you can `contest-generator`.
+
+Just define your plugins in
+[core-plugins.txt](/cmds/contest-generator/core-plugins.txt) with the format
```
core-plugins.txt -> .yml, here and in a few other places

insomniacslk replied:
whoops, thanks

insomniacslk replied:
done

tfg13 commented on cmds/contest-generator/README.md:
```diff
@@ -0,0 +1,63 @@
+the most common plugins.
+
+If you need to build `ConTest` with a custom set of plugins without maintaining
+your own `main.go` program, you can `contest-generator`.
```
I think there should be a verb somewhere here
```suggestion
your own `main.go` program, you use can `contest-generator`.
```

insomniacslk replied:
done

tfg13 commented on cmds/contest/main.go:
```diff
@@ -1,8 +1,6 @@
-// Copyright (c) Facebook, Inc. and its affiliates.
-//
-// This source code is licensed under the MIT license found in the
-// LICENSE file in the root directory of this source tree.
```
Won't the bot get angry about the missing copyright header?

insomniacslk replied:
it will, good catch

insomniacslk replied:
done

tfg13 commented on cmds/contest/server/server.go:
```diff
@@ -173,7 +244,7 @@ func ServerMain(cmd string, args []string, sigs <-chan os.Signal) error {
 
 	// spawn JobManager
-	listener := httplistener.NewHTTPListener(*flagListenAddr)
+	listener := httplistener.New(*flagListenAddr)
```
Should (maybe in a future version) the listener plugin be selectable in the generator as well?

insomniacslk replied:
yes, see my comment above. I am leaving it out for now as we don't have an interface for listeners yet, but I plan on adding it after this PR, and to adjust the generator to include that. But I wouldn't pack all of this in a single PR

insomniacslk replied:
I'm going to resolve this

tfg13 commented on cmds/contest/server/server.go:
```diff
@@ -61,7 +62,74 @@ func initFlags(cmd string) {
+	Factory job.ReporterFactory
+}
+
+// PluginConfig contains the configuration for all the plugins desired in a
+// server instance.
+type PluginConfig struct {
+	TargetManagers []TargetManagerConfig
+	TestFetchers   []TestFetcherConfig
+	TestSteps      []TestStepConfig
+	Reporters      []ReporterConfig
+}
+
+func registerPlugins(pluginRegistry *pluginregistry.PluginRegistry, pluginConfig *PluginConfig) error {
```
Almost everything in this file already exists in `cmds/plugins/plugins.go` (which is the wrong place now that I think of it)

insomniacslk replied:
I'll see what I can reuse, can't remember now why I recreated it

insomniacslk replied:
done

tfg13 commented on cmds/contest/server/server.go:
```diff
@@ -173,7 +244,7 @@ func ServerMain(cmd string, args []string, sigs <-chan os.Signal) error {
 	plugins.Init(pluginRegistry, ctx.Logger())
```
I think this initialises a hardcoded list of plugins again :/

insomniacslk replied:
good point, will look into it

insomniacslk replied:
Why do you think it's a hardcoded list? `pluginRegistry` is initialized with `pluginregistry.NewPluginRegistry` (which has an empty list of plugins), and only later populated in `server.Main` from the plugins written in the generated code

tfg13 replied:
Maybe I am reading it wrong, but won't `plugins.Init` end up here: https://github.com/facebookincubator/contest/blob/884d78e8b2a2f852997226df0ad0e14573f0d549/cmds/plugins/plugins.go#L66

And then load that hardcoded list of plugins?
Btw that thing also handles user functions for templates, they are basically plugins there.

insomniacslk replied:
ooh you're right. If I'm reading that correctly, we don't need to call `Init` at all, because we call `registerPlugins` with the plugins initialized in `main.go`. So just removing it should be enough?

insomniacslk replied:
I have removed it, and it seems it's only loading the plugins I expect. It also seems that the `cmds/plugins` package is not used anywhere else? Maybe it could be removed altogether?

insomniacslk replied:
except in `db/rdbms/migration/0002_migrate_descriptor_to_extended_descriptor.go` but if I understand correctly it's OK to remove that file, and who needs migrations will check out earlier commits?

tfg13 reviewed (changes requested):
See comments, I think this duplicates a lot of code and initialises some plugins twice.

The generator code lgtm.

Oh, and the commit contains a 15mb binary blob (server), please remove :)

insomniacslk commented:
Addressed feedback from Deomid, and added validation / duplicate checks. Will address Tobias's feedback in the next iteration

rojer9-fb commented on cmds/contest-generator/contest.go.template:
```diff
@@ -0,0 +1,78 @@
+{{/* This file is the template source. The following comment obviously doesn't apply here */ -}}
+// This is a generated file, edits should be made in the corresponding source
+// file, and this file regenerated using
+//   `contest-generator --from core-plugins.txt`.
```
this is a template that will be used with other configs, so the command will be wrong for uses other than core-plugins.
i think the template should have a variable, like {{ .CommandLine }}

insomniacslk replied:
I thought of that comment just as an example, not as the actual command. Maybe I can replace it with "--from your-plugins-configuration.yml" to be more clear?

insomniacslk replied:
changed, let me know if that's OK

insomniacslk replied:
ah I see. The generic name is also incorrect for `cmds/contest/main.go`

rojer9-fb replied:
well, now it won't match anyone's config :)
i was thinking of something along the lines of:

```
// Generated with:
//  { .CommandLine }}
```
with
```
  CommandLine: strings.Join(or.Args, " ")
```

or some such

insomniacslk replied:
done, with filename passed to the templaet

insomniacslk replied:
@rojer do you think we need the full command line, or a basic `contest-generator --from $filename` is enough? For example, if I run it with `go run` to regenerate cmds/contest/main.go,  it will look like
```
/tmp/somewhere/contest-generator -o ../contest/main.go
```
which is less useful than
```
contest-generator --from core-plugins.yml
```
(the plugins config file is not part of the command line, it's taken from the defaults)

insomniacslk commented:
I need to address more feedback, and add testing

rojer9-fb commented:
LGTM, good to go once Tobias is also happy

insomniacslk commented:
split config functions and types into `config.go`

insomniacslk commented:
will add tests for configuration objects after lunch

insomniacslk commented:
Tests added. @tfg13 PTAL

tfg13 commented on cmds/contest/server/server.go:
```diff
@@ -1,12 +1,8 @@
-// Copyright (c) Facebook, Inc. and its affiliates.
-//
-// This source code is licensed under the MIT license found in the
```
copyright header deleted here as well

insomniacslk replied:
done

insomniacslk commented:
fixed tests

mimir-d commented:
would you say that adding a build.sh script to do all of what the readme expects the user do to (go build stuff) is helpful?
I'm thinking since this PR has the semantics of providing a good to use binary with contest, maybe that goes the extra step for a turn-key solution.

insomniacslk commented:
> would you say that adding a build.sh script to do all of what the readme expects the user do to (go build stuff) is helpful?
> I'm thinking since this PR has the semantics of providing a good to use binary with contest, maybe that goes the extra step for a turn-key solution.

Not entirely sure if I got your point, but if you are referring to simplifying the process of building ConTest with contest-generator, there is already a pre-generated `main.go` under `cmds/contest`. I think a build script would only add more layers and confusion, and if the user needs to customize the build, they still need to edit a text file and call `contest-generator`, which is standard process for Go programs.

At the same time, this reminded me that `contest-generator` needs to be called by GitHub Actions, and rebuild and validate `cmds/contest/main.go`, which I'm going do to now

tfg13 commented on cmds/contest/server/server.go:
```diff
@@ -170,10 +223,8 @@ func ServerMain(cmd string, args []string, sigs <-chan os.Signal) error {
 		log.Fatalf("Invalid target locker name %q", *flagTargetLocker)
 	}
 
-	plugins.Init(pluginRegistry, ctx.Logger())
```
ok, this fixes loading plugins multiple times, but now we don't load the user functions at all anymore

insomniacslk replied:
good point, I'll re-add user function loading

mimir-d commented:
Yes, I'm talking about simplifying and guiding the user. The thing I have in mind is to have some extra checks for the user (I'm pretty sure a lot of people don't read the manual), like build.sh would point you to $EDITOR for the yaml file, then calls the contest-generator and go build and says "your binary is here".

insomniacslk commented:
> Yes, I'm talking about simplifying and guiding the user. The thing I have in mind is to have some extra checks for the user (I'm pretty sure a lot of people don't read the manual), like build.sh would point you to $EDITOR for the yaml file, then calls the contest-generator and go build and says "your binary is here".

Honestly I think this is too far. For that kind of users we should just provide prebuilt binaries

insomniacslk commented:
since `cmds/plugins` was removed because redundant, but still used by the migration tools, I moved that code directly into the migration tool

insomniacslk commented:
OK this is finally ready for the final review. A recap of the changes:

* added contest-generator (cmds/contest-generator/*)
* generated cmds/contest/main.go using contest-generator, which includes a comment with the commands to regenerate it
* tested contest-generator in GitHub actions so that it generates and validates the new code, compares it to cmds/contest/main.go, and fails if they are different (i.e. if the user forgot to regenerate it upon changes)
* removed the now redundant cmds/plugins/ and merged the remaining bits into the only consumer (the migration tool)
* included user documentation for contest-generator (cmds/contest-generator/README.md)
* provided default configuration file, with example for import aliasing (cmds/contest-generator/core-plugins.yml)
* tested parseConfig (cmds/contest-generator/config_test.go)
* renamed httplistener.NewHTTPListener to httplistener.New, waiting to create a listener interface in the next PRs (so that listeners will be configurable too - added as TODO)
* added TODO to include also user functions into code generation
* renamed cmds/contest/server.ServerMain to cmds/contest/server.Main (redundant prefix)
*

tfg13 commented on cmds/plugins/plugins.go:
```diff
@@ -1,107 +0,0 @@
-	}
-
-	// user-defined function registration
-	testInitOnce.Do(func() {
```
I think the test fails because you removed this `Once`, so on the second server start it tries to register the user functions again (they are not in the plugin registry, but in a global variable (boo!), so they should fail with "duplicate user function"). See below why this is not visible in the output

insomniacslk replied:
what to say, thanks for debugging it for me! Fix is coming right away

insomniacslk replied:
fixed

tfg13 commented on tests/e2e/e2e_test.go:
```diff
@@ -104,7 +108,16 @@ func (ts *E2ETestSuite) startServer(extraArgs ...string) {
+			TestStepLoaders:    []test.TestStepLoader{cmd.Load, sleep.Load},
+			ReporterLoaders:    []job.ReporterLoader{targetsuccess.Load, noop.Load},
+		}
+		_ = server.Main(&pc, "contest", args, serverSigs)
```
You should not throw away the error here, this should have said "error duplicate user function" :)

insomniacslk replied:
right. And it was already ignored in the original code, I just added `_ = ` to make the linter happy

insomniacslk replied:
done, now the e2e tests are surfacing errors in the test report instead of ignoring them

insomniacslk commented:
tests are green, and also improved error surfacing
